### PR TITLE
Enable exceptions, update to optimized newlib, migrate to new toolchain

### PR DESCRIPTION
**Pull request by earlephilhower**
This is a superset of #4694 that enables exceptions in user and library code by rebuilding the stdlibc++ to enable them, with patches (see [esp-quick-toolchain](https://github.com/earlephilhower/esp-quick-toolchain) for the patches to GCC building required).

It takes the work @d-a-v started, combines it with @igrr's suggestions in the original thread, fixes whatever problems I found, and makes exception handling "just work" on the 8266.

The current cost is ~1.2KB of RAM and ~19K of ROM (plus whatever add'l exception handling blocks your app would contain).  Looking to reduce that <1KB by limiting the emergency exception area (at the cost of OOM exceptions possibly double-faulting, but probably better than wasting 1KB of RAM for all apps).

It works as well as I've tested it (which is minimal), but it's 99.9% libstdc++ and G++, so I hope there's little for me to have flubbed.

WIP to keep the good stuff that #4694 started going.

### CODE REVIEW

**Comment by earlephilhower** [comment]
Still got a way to go before it's done.  Debug builds seem to overflow, number of emergency OOM spaces needs to be tested, the libc replacement `atexit()` (which didn't show up on my own tests) needs explaining.

That said it seems stable enough to be ready for people who need/want exceptions to give it a shakedown and help find hidden issues beyond the ones I know about.

**Comment by earlephilhower** [comment]
The glib math routines seem to be ending up in IRAM with this patch, whereas `master` uses the ROM version.  Looks like libc configuration needs some work.

````
earle@server:~/Arduino/hardware/esp8266com/esp8266$ diff exc.lst  nox.lst 
1,4d0
< __adddf3
< __adddf3_aux
< __addsf3
< __addsf3_aux
32,34d27
< __divdf3
< __divdf3_aux
< __divdi3
37d29
< __divsi3
46d37
< __fixdfsi
48,49d38
< __fixunsdfsi
< __fixunssfsi
52,55d40
< __floatsidf
< __floatsisf
< __floatunsidf
< __floatunsisf
113,115d97
< __muldf3
< __muldf3_aux
< __muldi3
172,175d153
< __subdf3
< __subdf3_aux
< __subsf3
< __subsf3_aux
191d168
< __truncdfsf2
195d171
< __udivsi3
206,207d181
< __umodsi3
< __umulsidi3
````

File: cores/esp8266/core_esp8266_main.cpp [hunk 1]
```diff
@@ -134,13 +134,27 @@ static void loop_task(os_event_t *events) {
...
 static void do_global_ctors(void) {
+    static struct object ob;
+    __register_frame_info( __eh_frame, &ob );
+
     void (**p)(void) = &__init_array_end;
     while (p != &__init_array_start)
         (*--p)();
 }
 
+void __gnu_cxx::__verbose_terminate_handler ()
```
Comment (igrr) [inline cores/esp8266/core_esp8266_main.cpp]: Could also be avoided by passing `--disable-libstdcxx-verbose` to GCC configure script

Comment (earlephilhower) [reply cores/esp8266/core_esp8266_main.cpp]: Thanks.  It's already built that way, so this is redundant code.  Will drop.

**Comment by earlephilhower** [comment]
Before going non-wip, one or two things remain:

1. Exception strings in std::exception are showing up in .RODATA and eating RAM, move them to FLASH somehow
````
 3ffe8d70 4a8dfe3f d0322040 488dfe3f 7374643a  J..?.2 @H..?std:
 3ffe8d80 3a657863 65707469 6f6e0073 74643a3a  :exception.std::
 3ffe8d90 6261645f 65786365 7074696f 6e007374  bad_exception.st
 3ffe8da0 643a3a62 61645f61 6c6c6f63 0067656e  d::bad_alloc.gen
 3ffe8db0 65726963 00626164 5f66756e 6374696f  eric.bad_functio
 3ffe8dc0 6e5f6361 6c6c0062 61736963 5f737472  n_call.basic_str
 3ffe8dd0 696e673a 3a617400 62617369 635f7374  ing::at.basic_st
 3ffe8de0 72696e67 3a3a636f 70790062 61736963  ring::copy.basic
 3ffe8df0 5f737472 696e673a 3a636f6d 70617265  _string::compare
 3ffe8e00 00626173 69635f73 7472696e 673a3a5f  .basic_string::_
 3ffe8e10 535f6372 65617465 00626173 69635f73  S_create.basic_s
 3ffe8e20 7472696e 673a3a65 72617365 00626173  tring::erase.bas
...
````
2. Probably want to make this yet-another-menu-option, so bring back the dummy exception fcns in a `#ifndef __EXCEPTIONS` block and adjust build.txt.py accordingly.

**Comment by earlephilhower** [comment]
esp-quick-toolchain now does a full newlib install to the Arduion dirs and copies the required `.a`s from elsewhere as appropriate.

Testing looks good, so we're good to go AFAICT.

**Pull request merged**